engine/src/main/java/io/camunda/zeebe/engine/state/migration/VersionCompatibilityCheck.java

```diff
@@ -0,0 +1,121 @@
+  private VersionCompatibilityCheck() {}
+
+  /**
+   * Checks the compatibility of the current version with the previous Version
+   *
+   * @param previousVersion a string representation of the previous version or null if unknown
+   * @param currentVersion a string representation of the current version or null if unknown
+   * @return a tri-state object indicating the compatibility of the versions:
+   *     <ul>
+   *       <li>{@link Indeterminate} if either of the versions are not available or version is not a
+   *           valid semantic version.
+   *       <li>{@link Incompatible} if the current version is not compatible with the previous
+   *           version, e.g. a downgrade or a skipped minor version.
+   *       <li>{@link Compatible} if the current version is compatible with the previous version,
+   *           e.g. upgrade to the next minor version.
+   */
+  public static CheckResult check(final String previousVersion, final String currentVersion) {
+    if (previousVersion == null) {
+      return new Indeterminate.PreviousVersionUnknown(currentVersion);
+    }
+    if (currentVersion == null) {
+      return new Indeterminate.CurrentVersionUnknown(previousVersion);
+    }
+
+    final var parsedPreviousVersion = SemanticVersion.parse(previousVersion);
+    final var parsedCurrentVersion = SemanticVersion.parse(currentVersion);
+
+    if (parsedPreviousVersion.isEmpty()) {
+      return new Indeterminate.PreviousVersionInvalid(previousVersion, currentVersion);
+    }
+
+    if (parsedCurrentVersion.isEmpty()) {
+      return new Indeterminate.CurrentVersionInvalid(previousVersion, currentVersion);
+    }
+
+    final var previous = parsedPreviousVersion.get();
+    final var current = parsedCurrentVersion.get();
+
+    if (previous.compareTo(current) == 0) {
+      return new Compatible.SameVersion(current);
+    } else if (previous.compareTo(current) > 0) {
+      if (previous.major() > current.major()) {
+        return new Incompatible.MajorDowngrade(previous, current);
+      } else if (previous.minor() > current.minor()) {
+        return new Incompatible.MinorDowngrade(previous, current);
+      } else if (previous.patch() > current.patch()) {
+        return new Incompatible.PatchDowngrade(previous, current);
+      } else {
+        return new Incompatible.PreReleaseDowngrade(previous, current);
+      }
+    } else {
+      if (previous.major() < current.major()) {
+        return new Incompatible.MajorUpgrade(previous, current);
+      } else if (previous.minor() - current.minor() < -1) {
+        return new Incompatible.SkippedMinorVersion(previous, current);
+      } else if (previous.minor() < current.minor()) {
+        return new Compatible.MinorUpgrade(previous, current);
+      } else if (previous.patch() < current.patch()) {
+        return new Compatible.PatchUpgrade(previous, current);
+      } else {
+        return new Compatible.PreReleaseUpgrade(previous, current);
+      }
+    }
```

Comment:
I'm not super happy with this because it seems brittle and somewhat hard to read. I'd be happy to hear suggestions on how this can be improved!